### PR TITLE
Use the audit ID of a request for better correlation

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Use the audit ID of a request for better correlation:
- If an audit ID is not provided by the caller, generate one when we receive the request. This audit ID will be used to audit the request. 
- When we forward the request to the aggregated server, set the audit ID in the new request header. This allows audit logs from aggregated apiservers to be correlated with the kube-apiserver
- Use the audit ID in the current `tracer` we have
- Use the audit ID in `httplog`
- When a request panics, log an error with the audit ID.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		auditID := r.Header.Get(auditinternal.HeaderAuditID)
+		if len(auditID) == 0 {
+			auditID = newAuditIDFunc()
+		}
+
+		r = r.WithContext(request.WithAuditID(ctx, types.UID(auditID)))
```
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go:53]: The audit ID is stored in the request context here. When a caller specifies an audit ID, it will also be in the request header `Audit-ID`. 
Another alternative is to add the generated audit id to the request header. But we treat the `http.Request` object as immutable.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 2]
```diff
@@ -762,8 +763,10 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
 	handler = genericapifilters.WithWarningRecorder(handler)
 	handler = genericapifilters.WithCacheControl(handler)
 	handler = genericfilters.WithHSTS(handler, c.HSTSDirectives)
+	handler = httplog.WithLogging(handler, httplog.DefaultStacktracePred)
 	handler = genericapifilters.WithRequestReceivedTimestamp(handler)
 	handler = genericfilters.WithPanicRecovery(handler, c.RequestInfoResolver)
+	handler = genericapifilters.WithAuditID(handler)
```
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:769]: It's before `WithPanicRecovery` - this is so we can access the audit ID from the context when we log error message inside the `WithPanicRecovery`  handler. 
If `WithAuditID` panics, `net/http2` will catch it and then log the panic with stack trace.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
+/*
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go:1]: nit: auditid.go

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+		auditID := r.Header.Get(auditinternal.HeaderAuditID)
+		if len(auditID) == 0 {
+			auditID = newAuditIDFunc()
+		}
+
+		r = r.WithContext(request.WithAuditID(ctx, types.UID(auditID)))
+
+		// Should we echo the Audit-ID for all requests?
+		// w.Header().Set(auditinternal.HeaderAuditID, auditID)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go:56]: I remember some discussion about that long long ago. @soltysh @deads2k maybe have better memory

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go:53]: This goes into direction of tracing. It rings by secrurity bell: if you want to inject this, you can fake it to hide yourself, and/or to put 8kb of data in a number of places (logs, audit events).

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/audit_id.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	return WithValue(parent, auditIDKey, auditID)
+}
+
+// AuditIDFrom returns the value of the auditID key from the specified context.
+func AuditIDFrom(ctx context.Context) (types.UID, bool) {
+	info, ok := ctx.Value(auditIDKey).(types.UID)
+	return info, ok
+}
+
+func GetVerbURIAndAuditID(req *http.Request) []interface{} {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/request/audit_id.go:49]: what is this?

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:769]: sounds fine

File: staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go [hunk 6]
```diff
@@ -51,16 +50,15 @@ func WithPanicRecovery(handler http.Handler, resolver request.RequestInfoResolve
 			// This call can have different handlers, but the default chain rate limits. Call it after the metrics are updated
 			// in case the rate limit delays it.  If you outrun the rate for this one timed out requests, something has gone
 			// seriously wrong with your server, but generally having a logging signal for timeouts is useful.
-			runtime.HandleError(fmt.Errorf("timeout or abort while handling: %v %q", req.Method, req.URL.Path))
+			runtime.HandleError(fmt.Errorf("timeout or abort while handling: %s", request.GetVerbURIAndAuditIDString(req)))
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:53]: What do we buy with this untyped way of writing this?

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go [hunk 7]
```diff
@@ -102,16 +109,20 @@ func respLoggerFromContext(req *http.Request) *respLogger {
 	return nil
 }
 
-// newLogged turns a normal response writer into a logged response writer.
-func newLogged(req *http.Request, w http.ResponseWriter) *respLogger {
+func newLoggedWithStartTime(req *http.Request, w http.ResponseWriter, startTime time.Time) *respLogger {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:112]: unrelated refactor?

File: staging/src/k8s.io/client-go/rest/request.go [hunk 8]
```diff
@@ -152,6 +154,9 @@ func NewRequest(c *RESTClient) *Request {
 	case len(c.content.ContentType) > 0:
 		r.SetHeader("Accept", c.content.ContentType+", */*")
 	}
+
+	// generate an audit ID
+	r.SetHeader("Audit-ID", uuid.New().String())
```
Comment (sttts) [inline staging/src/k8s.io/client-go/rest/request.go:159]: I feel like this whole cross-component topic makes the change much bigger, i.e. crossing the line to need an enhancement.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go:56]: Have you checked the audit enhancement? (if that exists)

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit_test.go [hunk 9]
```diff
@@ -772,16 +773,16 @@ func TestAuditIDHttpHeader(t *testing.T) {
 		expectedHeader bool
 	}{
 		{
-			"no http header when there is no audit",
+			"http header when there is no audit",
 			"",
 			auditinternal.LevelNone,
-			false,
+			true,
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit_test.go:779]: so this explicitly changed, right? Is the argument that we want to have an id in any case, even when we don't write audit events?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit_test.go:779]: A comment about the high level reason would be helpful

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+		}
+
+		// if the user has specified a very long audit ID then we will use the first N characters
+		// the UUID is formatted as '47aa608a-8931-4573-bd42-e3e3bc912ff0' (36 chars)
+		// so let's ensure we retain at least the first 36 characters.
+		// Note: assuming Audit-ID header is in ASCII
+		if len(auditID) > 36 {
+			attachAuditIDToCtx = true
+			auditID = auditID[0:36]
+		}
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go:66]: this is new. How do we know we don't break anybody? Why 36 chars?

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/auditid.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+//
+// First, we check if the audit ID has been attached to the the request context.
+// If not, we retrieve it from the request writeToHeader.
+func AuditIDFrom(r *http.Request) (types.UID, bool) {
+	ctx := r.Context()
+	if auditID, ok := ctx.Value(auditIDKey).(types.UID); ok {
+		return auditID, ok
+	}
+
+	if auditID := r.Header.Get(auditinternal.HeaderAuditID); len(auditID) > 0 {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/request/auditid.go:53]: I find all the logic starting from here irritating in this context. Would prefer exactly one place where we have header logic.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go (hunk 10 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go:66]: - yes, since we are copying the audit ID into the request context I wanted to a put a defensive guard to avoid copying large amount of data into the context. 
- Also, printing a very large audit ID is going to make reading the log very cumbersome. The client can still search with the first 36 characters.
- I chose 36, because the formatted UID we generate today is 36 characters long. I am fine increasing it but would still like us to have a limit.
- Well client-go does not generate any audit ID today. I don't think it breaks anyone today. The caller can still search with the first N characters.

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/auditid.go (hunk 11 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/auditid.go:53]: @sttts This is an optimization - If the audit ID is in the request header and we can use it as is then avoid copying it to the context. do you want me to go back to `always copy the audit ID in the request context`?

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go (hunk 4 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go]: Going through the comments, it looks like the intention was to set it for all requests. So I think moving it here makes sense.

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/audit_id.go (hunk 5 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/audit_id.go]: I removed it and doing it inline now.

File: staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go (hunk 6 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go]: having a consistent key names (with structured logging). But I remove it and doing it inline now.

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go (hunk 7 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go]: Yes, will open a follow up PR. the httplogger should use the `RequestReceivedTimestamp` from the request context. Then we will have both audit and httplog print the same exact timestamp.

File: staging/src/k8s.io/client-go/rest/request.go (hunk 8 above)
Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/request.go]: yes, I removed the client-go changes. I will create a new KEP with the proposed client-go changes.

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/auditid.go (hunk 11 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/auditid.go]: @sttts I removed the optimization, the context always has an audit ID.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 12]
```diff
@@ -762,8 +763,10 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
 	handler = genericapifilters.WithWarningRecorder(handler)
 	handler = genericapifilters.WithCacheControl(handler)
 	handler = genericfilters.WithHSTS(handler, c.HSTSDirectives)
+	handler = httplog.WithLogging(handler, httplog.DefaultStacktracePred)
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:766]: why have we moved it here?

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go [hunk 13]
```diff
@@ -157,12 +158,14 @@ func (rl *respLogger) Addf(format string, data ...interface{}) {
 
 func (rl *respLogger) LogArgs() []interface{} {
 	latency := time.Since(rl.startTime)
+	auditID := request.GetAuditID(rl.req)
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:161]: it cannot be empty, right ?

File: staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go [hunk 14]
```diff
@@ -51,16 +50,15 @@ func WithPanicRecovery(handler http.Handler, resolver request.RequestInfoResolve
 			// This call can have different handlers, but the default chain rate limits. Call it after the metrics are updated
 			// in case the rate limit delays it.  If you outrun the rate for this one timed out requests, something has gone
 			// seriously wrong with your server, but generally having a logging signal for timeouts is useful.
-			runtime.HandleError(fmt.Errorf("timeout or abort while handling: %v %q", req.Method, req.URL.Path))
+			runtime.HandleError(fmt.Errorf("timeout or abort while handling: method=%v URI=%q audit-ID=%q", req.Method, req.RequestURI, request.GetAuditID(req)))
 			return
 		}
 		http.Error(w, "This request caused apiserver to panic. Look in the logs for details.", http.StatusInternalServerError)
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:56]: should we change the msg to contain the correlation ID as well? Users won't see it by default only with increased log level.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go (hunk 10 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go:63]: Does it mean the input hasn't been sanitized before? so clients could send potentially very large inputs.
AFAIK the HTTP spec doesn't specify a limit for Headers.

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go]: for now, I am saving the audit ID as is in the context. But when we are logging (tracing and httplog) I am only showing the first N(64) character.

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 12 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: good catch, this was an artifact for another follow up PR. I remove it.

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go (hunk 13 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go]: if `WithAuditID` is wired, it cannot be empty.

File: staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go (hunk 14 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:57]: well, the audit ID is set in the response header. Also, for when net/http catches the panic it will close the connection or reset the stream for http2. So the user doesn't see the http response, I think.

**Comment by tkashem** [comment]
Hi @liggitt can you please take a look when you have some time? 

/assign @liggitt

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/helpers.go [hunk 15]
```diff
@@ -73,3 +74,17 @@ func (lazy *lazyAccept) String() string {
 
 	return "unknown"
 }
+
+// lazyAuditID implements String() string and it will retrieve the audit ID
+// (associated with the request) lazily only when required.
```
Comment (soltysh) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/helpers.go:79]: Suggestion:

lazyAuditID implements Stringer interface to lazily retrieve the audit ID associated with the request

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go [hunk 16]
```diff
@@ -201,6 +202,12 @@ func newRequestForProxy(location *url.URL, req *http.Request) (*http.Request, co
 	newReq.URL = location
 	newReq.Host = location.Host
 
+	// If the original request has an audit ID, let's make sure we propagate this
+	// to the aggregated server.
+	if auditID, found := genericapirequest.AuditIDFrom(req.Context()); found {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:207]: have you checked `newReq.Header = utilnet.CloneHeader(req.Header)` ?

`req` should already have the `AuditID` header set by the filter

File: staging/src/k8s.io/apiserver/pkg/audit/request.go [hunk 17]
```diff
@@ -52,14 +53,11 @@ func NewEventFromRequest(req *http.Request, requestReceivedTimestamp time.Time,
...
 	}
 
-	// prefer the id from the headers. If not available, create a new one.
-	// TODO(audit): do we want to forbid the header for non-front-proxy users?
-	ids := req.Header.Get(auditinternal.HeaderAuditID)
-	if ids != "" {
-		ev.AuditID = types.UID(ids)
-	} else {
-		ev.AuditID = types.UID(uuid.New().String())
+	auditID, found := request.AuditIDFrom(req.Context())
```
[the remark this answers is not in the document]
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/audit/request.go:56]: yes, it is set when the builder chain has `WithAuditID` filter. Here, it is just handling an error case.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+// a. If the caller does not specify a value for Audit-ID in the request header, we generate a new audit ID
+// b. We echo the Audit-ID value to the caller via the response Header 'Audit-ID'.
+func WithAuditID(handler http.Handler) http.Handler {
+	return withAuditID(handler, func() string {
+		return uuid.New().String()
+	})
+}
+
+func withAuditID(handler http.Handler, newAuditIDFunc func() string) http.Handler {
+	if newAuditIDFunc == nil {
```
[the remark this answers is not in the document]
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go:40]: it should not be `nil`, here it is just handling an error path. Maybe I can print an error log message here.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go (hunk 16 above)
Comment (tkashem) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:207]: If the user specify an audit ID only then `req` will already have the `Audit-ID` header. If we generate an audit ID then we save it in the context. 
if the context already has an audit ID then it has higher precedence, and we want to set it to the new `req`.

Comment (p0lyn0mial) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:207]: okay, makes sense, thx.

File: staging/src/k8s.io/apiserver/pkg/audit/request.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/audit/request.go:56]: could you point me to the error case?

[the remark this answers is not in the document]
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/audit/request.go:56]: by `error case` i meant handling the case where `found` is false, instead of doing `auditID, _ := request.AuditIDFrom(req.Context())`
